In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv("data/boston.csv")
data.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

Dữ liệu BOSTON: Xây dựng mô hình dự báo medv theo age và lstat. 
Biết rằng cần chuẩn hóa dữ liệu StandardScaler cho dữ liệu input trên tập training. Dữ liệu sẽ được cắt theo tỉ lệ train:test là 90:10 với random state là 10


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score , mean_absolute_error

# Defining the features and target variable
X = data[['AGE', 'LSTAT']]
y = data['MEDV']

# Splitting the data into training and test sets with a ratio of 90:10 and random state of 10
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=10)

# Standardizing the training data using StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Building a Linear Regression model
model = LinearRegression()
model.fit(X_train_scaled, y_train)

# Making predictions on the test set
y_pred = model.predict(X_test_scaled)

# Evaluating the model using Mean Squared Error (MSE) and R-squared (R2) metrics
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)

mse, r2


(np.float64(49.312606628201), 0.5019088013216607)

**Dự báo giá trị `medv` khi biết `age` = 37.5 và `lstat` = 4.65**

a. 28.45479863

b. 0.35633368

c. 25.45479863

d. 29.87462092

e. 23.45479863

f. 22.45479863

g. 29.45479863

h. 26.45479863

i. 27.45479863

j. 24.45479863

k. 0.08837772


In [4]:
# Using the trained model to predict MEDV for given AGE and LSTAT values

# Input values for prediction
input_data = [[37.5, 4.65]]

# Scaling the input data based on the training scaler
input_data_scaled = scaler.transform(input_data)

# Predicting MEDV using the model
medv_prediction = model.predict(input_data_scaled)[0]
medv_prediction


c:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


np.float64(29.454798629362067)

**Phần trăm sự thay đổi của biến phụ thuộc được giải thích bởi các biến độc lập**

a. 0.5630979953818331

b. 0.4069676937197577

c. 0.5019088013216606

d. 0.1012807931898763

e. 0.2012807931898763

f. 0.6012807931898763

g. 6.264737108611292

h. 0.5480883767100382


In [5]:
r2

0.5019088013216607

**Hệ số góc của biến độc lập `age` trong mô hình huấn luyện**

a. 0.74746824

b. 2.04746824

c. 5.58919129

d. 0.84338848

e. 0.64746824

f. 31.26170026

g. 1.04746824

h. 0.94746824

i. -38.23018134

j. 3.04746824

k. 4.04746824


In [6]:
age_coefficient = model.coef_[0]
age_coefficient

np.float64(0.8433884781639227)

**Dòng dữ liệu đầu tiên của tập dữ liệu đầu ra trên tập `train`**

a. 33.8

b. 24.3

c. 15.2

d. 16.1

e. 24.2

f. 12.5

g. 22.6

h. 43.8

i. 23.4

j. 19.3

k. 22



In [7]:
# Displaying the first value of the target variable (MEDV) in the training set
first_train_output = y_train.iloc[0]
first_train_output


np.float64(12.5)

**Dòng dữ liệu cuối cùng của tập dữ liệu đầu vào trên tập `train`**

a. [-0.21856566, -0.34247973]

b. [0.47029208, -0.42936116]

c. [0.73699637, -0.86530163]

d. [-1.1263776, -1.27320734]

e. [-0.01688776, -0.14060317]

f. [0.45251179, 0.00237409]

g. [0.51493306, 0.2872276]

h. [0.95056643, 0.29025424]

i. [-0.12001342, 0.13413745]

j. [0.9752522, 0.94153844]

k. [-1.09437309, -0.00183112]


In [8]:
# Displaying the last row of the input training set after scaling
last_train_input = X_train_scaled[-1]
last_train_input


array([-0.21856566, -0.34247973])

**Hệ số tự do trong mô hình huấn luyện**

a. 26.3459616

b. 29.3459616

c. 31.26170026

d. 35.26170026

e. 30.26170026

f. 28.3459616

g. 27.3459616

h. 23.3459616

i. 24.3459616

j. 22.12483516

k. 25.3459616


In [9]:
# Retrieving the intercept (constant term) of the trained linear model
intercept = model.intercept_
intercept


np.float64(22.12483516483517)

**Dòng đầu tiên sau khi chuẩn hóa dữ liệu trên tập `train`**

a. [0.52935118, 0.29207022]

b. [0.92481977, 0.49424939]

c. [0.97880825, 0.91070021]

d. [-0.80988851, -1.36151682]

e. [-0.35115713, -1.04332233]

f. [-0.51117971, -1.02650148]

g. [-0.26581176, -1.2087274]

h. [0.36716642, -0.49243937]

i. [-0.07022862, -0.03126762]

j. [0.64160659, 0.09836562]

k. [0.96775636, 0.91299676]


In [10]:
# Displaying the first row of the input training set after scaling
first_train_input_scaled = X_train_scaled[0]
first_train_input_scaled


array([0.96775636, 0.91299676])

**Giá trị MAPE của mô hình trên tập kiểm thử**

a. 5.185117026484556

b. 4.5918745715026485

c. 39.24693104001137

d. 0.2547803169337023

e. 0.5019088013216606

f. 7.02229354471892

g. 0.2012807931898763

h. 49.31260662820101


In [11]:
import numpy as np

# Calculating Mean Absolute Percentage Error (MAPE) for the model on the test set
mape = np.mean(np.abs((y_test - y_pred) / y_test)) * 100
mape


np.float64(20.12807931898763)

**Kích thước của tập dữ liệu `train`**

a. (401, 2)

b. (458, 2)

c. (404, 2)

d. (459, 2)

e. (406, 2)

f. (457, 2)

g. (456, 2)

h. (455, 2)


In [12]:
# Displaying the shape of the training dataset
train_shape = X_train.shape
train_shape


(455, 2)